### PR TITLE
Banshee wfi

### PR DESCRIPTION
**Pull request by SamuelRiedel**
The RISC-V `wfi` instruction currently stops the simulation of a core. However, this operation is also commonly used to put cores to sleep, for example at barriers, or queue locks. To support this behavior I suggest the following:
- We have global variables to count how many cores are asleep and which cores should be woken up.
- Instead of terminating the thread immediately when getting a `wfi`, the thread instead checks if all cores are asleep. If this is the case, terminate the simulation.
- If some cores are still awake, the thread should yield. When it becomes active again, it should check for a wake-up signal or whether all cores are asleep.

Todo:
- [x] Discuss the behavior
- [x] Make sure the simulation is not ended prematurely. For example, if the only core awake triggers the next core in a queue lock and then goes to sleep itself, all cores are asleep, until the next thread sees its wakes-up trigger. We have to make sure none of the cores finish the simulation in such a situation.
- [x] Discuss the concept of time spent sleeping. Currently, we only count the retired instructions. The time a core spent sleeping is therefore always just the one `wfi` cycle. A possibility could be that the core waking up inherits the time of the core sending the trigger if this time is bigger. However, adding a notion of time should be discussed in a different PR, in my opinion.
- [x] Add unittest

This PR relies on #34.

### CODE REVIEW

**Comment by SamuelRiedel** [comment]
@fabianschuiki What do you think?

File: sw/banshee/src/engine.rs [hunk 1]
```diff
@@ -316,6 +316,12 @@ impl Engine {
             .map(|_| AtomicUsize::new(0))
             .collect();
 
+        // Allocate global variables to keep track of sleeping cores.
+        let num_sleep = AtomicUsize::new(0);
+        let wake_up: Vec<_> = (0..self.num_clusters * self.num_cores)
+            .map(|_| AtomicUsize::new(0))
+            .collect();
+
```
Comment (fabianschuiki) [inline sw/banshee/src/engine.rs:324]: This seems like a nice way to track sleeping cores :+1:

File: sw/banshee/src/engine.rs [hunk 2]
```diff
@@ -476,7 +492,20 @@ impl<'a, 'b> Cpu<'a, 'b> {
...
+                if value as i32 == -1 {
+                    self.num_sleep
+                        .fetch_sub(self.wake_up.len(), Ordering::Relaxed);
+                    for x in self.wake_up {
+                        x.fetch_add(1, Ordering::Relaxed);
+                    }
+                } else if (value as usize) < self.wake_up.len() {
+                    self.num_sleep.fetch_sub(1, Ordering::Relaxed);
+                    self.wake_up[value as usize].fetch_add(1, Ordering::Relaxed);
+                }
```
Comment (fabianschuiki) [inline sw/banshee/src/engine.rs:506]: Neat!

File: sw/banshee/src/tran.rs [hunk 3]
```diff
@@ -1903,7 +1903,21 @@ impl<'a> InstructionTranslator<'a> {
...
+                    LLVMCreateBasicBlockInContext(self.section.engine.context, NONAME);
+                let bb_wake_up = LLVMCreateBasicBlockInContext(self.section.engine.context, NONAME);
+                LLVMInsertExistingBasicBlockAfterInsertBlock(self.builder, bb_terminate);
+                LLVMInsertExistingBasicBlockAfterInsertBlock(self.builder, bb_wake_up);
+                LLVMBuildCondBr(self.builder, terminate, bb_terminate, bb_wake_up);
+                // Terminate
+                LLVMPositionBuilderAtEnd(self.builder, bb_terminate);
+                LLVMBuildRetVoid(self.builder);
+                // Continue
+                LLVMPositionBuilderAtEnd(self.builder, bb_wake_up);
```
Comment (fabianschuiki) [inline sw/banshee/src/tran.rs:1920]: Much better 😉

File: sw/banshee/tests/wfi.c [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+        *x = 1;
+        *wake_up = (core_id + 1) % core_num;
+    } else {
+        asm volatile("wfi");
+        *x += 1;
+        *wake_up = (core_id + 1) % core_num;
+    }
+    asm volatile("wfi");
+    return core_id == 0 ? core_num -= *x : 0;
+}
```
Comment (fabianschuiki) [inline sw/banshee/tests/wfi.c:21]: Nice test!

**Review by fabianschuiki — APPROVED** [review]
LGTM!

File: sw/banshee/src/engine.rs (hunk 2 above)
Comment (SamuelRiedel) [reply sw/banshee/src/engine.rs:506]: Thanks :)

File: sw/banshee/tests/wfi.c (hunk 4 above)
Comment (SamuelRiedel) [reply sw/banshee/tests/wfi.c:21]: Perfect. Then we can merge :rocket:

**Pull request merged**